File: 241114/그래프 탐색/graph-traversal.js
const readline = require('readline');

// 입력을 처리하기 위한 인터페이스 생성
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// 입력 저장을 위한 배열
let input = [];

// 한 줄씩 입력을 배열에 저장
rl.on('line', (line) => {
    input.push(line);
}).on('close', () => {
    // 첫 번째 줄에서 N과 M을 파싱
    const [N, M] = input[0].split(' ').map(Number);
    
    // 두 번째 줄부터는 간선 정보 파싱
    const edges = input.slice(1).map(line => line.split(' ').map(Number));
    
    // 함수에 N, M, edges를 전달하여 실행
    console.log(countReachableNodes(N, M, edges));
    
    process.exit();
});

// 그래프에서 도달할 수 있는 서로 다른 정점의 수를 계산하는 함수
function countReachableNodes(N, M, edges) {
    // 1. 그래프를 인접 리스트로 표현
    const graph = Array.from({ length: N + 1 }, () => []);
    
    // 2. 간선 정보를 통해 그래프를 구축
    edges.forEach(([x, y]) => {
        graph[x].push(y);
        graph[y].push(x);
    });
    
    // 3. 방문 여부를 기록할 배열
    const visited = Array(N + 1).fill(false);
    
    // 4. DFS 탐색 함수 정의
    function dfs(node) {
        visited[node] = true;
        for (let neighbor of graph[node]) {
            if (!visited[neighbor]) {
                dfs(neighbor);
            }
        }
    }
    
    // 5. 1번 정점에서 DFS 시작
    dfs(1);
    
    // 6. 1번 정점을 제외한 방문한 정점의 개수를 카운트
    let count = 0;
    for (let i = 2; i <= N; i++) {
        if (visited[i]) {
            count++;
        }
    }
    
    return count;
}

File: 250919/그래프 탐색/graph-traversal.java
import java.util.*;

public class Main {
    static ArrayList<Integer>[] graph; // 인접 리스트
    static boolean[] visited;          // 방문 체크
    static int count = 0;              // 도달 가능한 정점 수

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt(); // 정점 수
        int m = sc.nextInt(); // 간선 수

        graph = new ArrayList[n + 1];
        visited = new boolean[n + 1];

        // 인접 리스트 초기화
        for (int i = 1; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }

        // 간선 입력 (양방향)
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            graph[u].add(v);
            graph[v].add(u);
        }

        // DFS 시작 (1번 정점에서 출발)
        dfs(1);

        // 시작점(1번)은 제외해야 하므로 -1
        System.out.println(count - 1);
    }

    static void dfs(int node) {
        visited[node] = true;
        count++;

        for (int next : graph[node]) {
            if (!visited[next]) {
                dfs(next);
            }
        }
    }
}
